config/decoder.go

```diff
@@ -205,9 +212,16 @@ func (d *decoder) mapping(childKey string, value reflect.Value, def string) erro
 	// child yamlNode parsed from yaml file is of type map[interface{}]interface{}
 	// type casting here makes sure that we are iterating over a parsed map.
 	if v, ok := val.(map[interface{}]interface{}); ok {
-		childKey += _separator
+		childKey = addSeparator(childKey)
+
 		for key := range v {
 			mapKey := childKey + fmt.Sprintf("%v", key)
+
+			// Skip empty keys, because lookups in provider can toggle infinite loops/spill internals.
+			if mapKey == "" {
+				continue
```

Comment:
minor: Instead of skip/continue checking for not empty is more clear. we can combine with line 218.